### PR TITLE
Add metric QC function

### PR DESCRIPTION
**Pull request by lahuuki**
Adding a function to aid/standardize the QC process of new `spe` objects based off of work in [Habenula_Visium](https://github.com/LieberInstitute/Habenula_Visium/blob/7fc73e8cb403bcaa035f2451602db3fc0f38b103/code/02_build_spe/02_exploratory_QCs.R#L188-L295), [spatial_NAc](https://github.com/LieberInstitute/spatial_NAc/blob/56c67e7c0421e6dcf71fe2db6ef2c7b62ed15e58/code/05_harmony_BayesSpace/02-compute_QC_metrics.R#L69-L128), and [Visium_SPG_AD](https://github.com/LieberInstitute/Visium_SPG_AD/blob/e13e1a3604aa2dc43d76fa6923541ef403b0b13d/code/07_spot_qc/05_qc_plots_revision.R#L89-L152).  

Utilizes `scater::isOutlier` on three metrics `log2(sum_umi)`, `log2(sum_gene)`, and `expr_chrM_ratio*100`. 
Added functionality for handling spots `in_tissue=FALSE` .

### CODE REVIEW

File: R/metrics_qc.R [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+#'  @importFrom dplyr group_by summarize left_join select mutate
+#'  @importFrom SummarizedExperiment colData
+#'  @importFrom scater isOutlier
+metrics_qc <- function(spe) {
+    stopifnot("in_tissue" %in% colnames(colData(spe)))
+    stopifnot("sum_umi" %in% colnames(colData(spe)))
+    stopifnot("sum_gene" %in% colnames(colData(spe)))
+    stopifnot("expr_chrM_ratio" %in% colnames(colData(spe)))
+
+    spe_in <- spe[, spe$in_tissue]
```
Comment (lcolladotor) [inline R/metrics_qc.R]: Sometimes `spe$in_tissue` is a `factor` or even a `character`, so this might not work in general.

File: R/metrics_qc.R [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        subsets_Mito_percent = spe_in$expr_chrM_ratio * 100,
+        sample_id = spe_in$sample_id
+    )
+
+    qcfilter <- data.frame(
+        low_lib_size = scater::isOutlier(qc_df$log2sum, type = "lower", log = TRUE, batch = qc_df$sample_id),
+        low_n_features = scater::isOutlier(qc_df$log2detected, type = "lower", log = TRUE, batch = qc_df$sample_id),
+        high_subsets_Mito_percent = scater::isOutlier(qc_df$subsets_Mito_percent, type = "higher", batch = qc_df$sample_id)
+    ) |>
+        dplyr::mutate(discard = (low_lib_size | low_n_features) | high_subsets_Mito_percent)
```
Comment (lcolladotor) [inline R/metrics_qc.R]: I'd rather not add `dplyr` as a dependency just for this `mutate()` function call.

File: R/metrics_qc.R [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    spe$scran_high_subsets_Mito_percent[which(spe$in_tissue)] <-
+        qcfilter$high_subsets_Mito_percent
+    spe$scran_high_subsets_Mito_percent <-
+        factor(spe$scran_high_subsets_Mito_percent, levels = c("TRUE", "FALSE"))
+
+    ## Find edge spots
+    spot_coords <- colData(spe_in) |>
+        as.data.frame() |>
+        select(in_tissue, sample_id, array_row, array_col) |>
+        group_by(sample_id, array_row) |>
```
Comment (lcolladotor) [inline R/metrics_qc.R]: Ok, I see more `dplyr` function calls here. So maybe we can end up adding it then.

If that's the case, it has to be added to the `DESCRIPTION` file too.

File: R/metrics_qc.R (hunk 1 above)
Comment (lcolladotor) [reply R/metrics_qc.R]: You might want to include an explicit `spe$in_tissue <- as.logical(spe$in_tissue)` call. It'll error out automatically on cases where a factor/character can't be casted into a logical vector.

File: R/metrics_qc.R [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+#'     scater::plotColData(spe_qc[, spe_qc$in_tissue], x = "sample_id", y = "sum_umi", colour_by = "scran_low_lib_size")
+#'
+#'     vis_clus(spe_qc, sampleid = "151507", clustervar = "scran_discard")
+#'     vis_clus(spe_qc, sampleid = "151507", clustervar = "scran_low_lib_size_edge")
+#' }
+#' #'
+#'  @importFrom dplyr group_by summarize left_join select mutate
+#'  @importFrom SummarizedExperiment colData
+#'  @importFrom scater isOutlier
+metrics_qc <- function(spe) {
```
Comment (lcolladotor) [inline R/metrics_qc.R]: `metrics_qc()` was the name of the function I made, but maybe we should change it to a more understandable name. Typically using verbs helps for this. For example, `add_metrics_qc()` or `compute_qc_metrics()`.

File: R/metrics_qc.R [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+#'
+#'     scater::plotColData(spe_qc[, spe_qc$in_tissue], x = "sample_id", y = "sum_umi", colour_by = "scran_low_lib_size")
+#'
+#'     vis_clus(spe_qc, sampleid = "151507", clustervar = "scran_discard")
+#'     vis_clus(spe_qc, sampleid = "151507", clustervar = "scran_low_lib_size_edge")
+#' }
+#' #'
+#'  @importFrom dplyr group_by summarize left_join select mutate
+#'  @importFrom SummarizedExperiment colData
+#'  @importFrom scater isOutlier
```
Comment (lcolladotor) [inline R/metrics_qc.R]: Looks like `scater` has to be added to the `DESCRIPTION` file. Use `usethis::use_package()` for this.

**Review by lcolladotor — COMMENTED** [review]
Initial PR comments

File: R/metrics_qc.R (hunk 4 above)
Comment (lahuuki) [reply R/metrics_qc.R]: How about `add_qc_metrics()` ? 
`add_spatial_qc_metrics()` would be the most verbose but gets clunky

Comment (lcolladotor) [reply R/metrics_qc.R]: Sure, sounds good!

File: R/metrics_qc.R (hunk 1 above)
Comment (lahuuki) [reply R/metrics_qc.R]: Added this check!

File: R/metrics_qc.R (hunk 5 above)
Comment (lahuuki) [reply R/metrics_qc.R]: Fixed

File: R/metrics_qc.R (hunk 4 above)
Comment (lahuuki) [reply R/metrics_qc.R]: I changed the name

**Review by lcolladotor — COMMENTED** [review]
Let's merge it =)

**Pull request merged**